### DISCUSSION TITLE
Implementing Active Directory

### DISCUSSION THREAD

**Discussion by laredoza** [elsa-workflows/elsa-core#5611] [General]
Hi,

I'd appreciate some guidance with implementing ClientId/ClientSecret authentication using AzureAD(Entra Id).

I've split Elsa into 2 projects, an API Server and a Blazor Backend.

Api Server Program.cs

```cs
builder.Services.AddControllers(options =>
			{
				options.EnableEndpointRouting = false;
			});

builder.Services.AddAuthentication(options =>
{
    options.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
})
.AddJwtBearer(options =>
{
    options.Authority = $"{builder.Configuration["AzureAd:Instance"]}{builder.Configuration["AzureAd:TenantId"]}";
    options.Audience = builder.Configuration["AzureAd:ClientId"];
    options.TokenValidationParameters = new TokenValidationParameters
    {
        ValidateIssuer = true,
        ValidIssuer = $"{builder.Configuration["AzureAd:Instance"]}{builder.Configuration["AzureAd:TenantId"]}",
        ValidateLifetime = true,
        ClockSkew = TimeSpan.Zero
    };
    options.Events = new JwtBearerEvents
    {
        OnAuthenticationFailed = context =>
        {
            Console.WriteLine("Authentication failed.");
            return Task.CompletedTask;
        },
        OnTokenValidated = context =>
        {
            Console.WriteLine("Token validated.");
            return Task.CompletedTask;
        },
    };
});

builder.Services.AddElsa(elsa =>
	{
		string connectionString = builder.Configuration.GetConnectionString("SCHEDULEDB");
		// string connectionString = builder.Configuration.GetConnectionString("SCHEDULEDB");
		elsa.UseWorkflowManagement(management => management.UseEntityFrameworkCore(ef => ef.UseSqlServer(connectionString)));
		elsa.UseWorkflowRuntime(runtime => runtime.UseEntityFrameworkCore(ef => ef.UseSqlServer(connectionString)));
		elsa.UseWorkflowsApi();

		// Setup a SignalR hub for real-time updates from the server.
		elsa.UseRealTimeWorkflows();

		// Enable C# workflow expressions
		elsa.UseCSharp();
		elsa.UseJavaScript();
		elsa.UseLiquid();
	});

// Configure web application's middleware pipeline.
app.UseCors();
app.UseStaticFiles();

app.UseRouting(); // Required for SignalR.
app.UseAuthentication();
app.UseAuthorization();
app.UseWorkflowsApi(); // Use Elsa API endpoints.
app.UseWorkflows(); // Use Elsa middleware to handle HTTP requests mapped to HTTP Endpoint activities.
app.UseWorkflowsSignalRHubs(); // Optional SignalR integration. Elsa Studio uses SignalR to receive real-time updates from the server. 

app.UseEndpoints(endpoints =>
{
    endpoints.MapControllers();
});

app.Run();
```
 
I've added a controller with an endpoint that I can call with the Bearer token I created using a ClientId/ClientSecret combination.

`
curl --location 'http://localhost:5001/test/test' \
--header 'Authorization: ••••••'
`

This works as expected.

When I try and call an admin endpoint I get a forbidden message.

`
curl --location 'http://localhost:5001/elsa/api/workflow-definitions?versionOptions=Latest&Page=0&PageSize=10&SearchTerm=&IsSystem=False&OrderDirection=Ascending' \
--header 'Authorization: ••••••'
`

From the looks of it, I need specific permissions. What do I need to do to make this work? How can I assign an approle to my ApplicationRegistration so that it works for this? I'd appreciate being pointed in the correct direction.

**Comment by glime-ai** [comment]
To call an admin endpoint on your Elsa API Server with Azure AD authentication, you need to ensure that the token you are using has the necessary roles or scopes to access the admin resources. Here's what you need to do:

1. **Define App Roles in Azure AD**: In your Azure AD application registration, define app roles that correspond to the permissions required by your API. For example, you might create an app role called `ElsaAdmin` with the necessary permissions to access admin endpoints.

2. **Assign App Roles to Service Principal**: Assign the defined app roles to the service principal (application) that represents your client application in Azure AD. This step effectively grants the client application the permissions associated with the role.

3. **Acquire Token with Role**: When acquiring a token using the ClientId/ClientSecret, ensure that the token request includes the app role. This will result in a token that contains the role claim.

4. **Authorize Based on Role in API**: In your API Server's `Startup.cs` or `Program.cs`, configure the authorization policies to require the `ElsaAdmin` role for accessing admin endpoints. You can use the `RequireRole` method when defining authorization policies.

Here's an example of how you might configure authorization policies in your API Server:

```csharp
builder.Services.AddAuthorization(options =>
{
    options.AddPolicy("RequireElsaAdminRole", policy =>
        policy.RequireRole("ElsaAdmin"));
});
```

And then apply the policy to your admin endpoints:

```csharp
[Authorize(Policy = "RequireElsaAdminRole")]
public class AdminController : ControllerBase
{
    // Your admin actions
}
```

5. **Test the Secured Endpoint**: Use the acquired token to call the secured admin endpoint. The token should now contain the role claim, and the API should authorize the request if the role matches.

Please note that the exact steps for defining app roles and assigning them to a service principal can vary depending on your Azure AD setup and whether you are using the Azure portal or Azure AD PowerShell/CLI for configuration.

If you need further assistance with the specific steps in Azure AD, you can refer to the official Microsoft documentation on app roles and role-based access control (RBAC) in Azure AD.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by laredoza** [comment]
I don't think this addresses my issue in Elsa3,  as the admin section is created with FastEndpoints and I don't have access to it using a nuget package.  I'm assuming I need to create my own DefaultAuthenticationFeature. I'd like some guidance on which providers I need to create/update?

I know there is the APIKey but I'd prefer to manage everything in one place if possible.

**Comment by zuntio** [comment]
Please see my solution [here](https://github.com/elsa-workflows/elsa-core/discussions/4814#discussioncomment-9398844). Not accepted as answer or anything but at least it works.

Any feedback or improvements are appreciated.

**Reply by laredoza** [reply]
Thanks, I'll give it a go tomorrow.

**Reply by laredoza** [reply]
@zuntio I've tested a postman call to the API after I implemented ElsaPermissionTransformation and it worked perfectly. Thanks, you saved me a huge amount of time. This should be added to the documentation. 

I'll try and implement the Blazor part soon.

**Reply by laredoza** [reply]
@zuntio I've managed to get the Blazor side of things working following your guide.

The only change I had to make was:
```cs
.AddSingleton<IBackendJwtHandler, ElsaBackendJwtHandler>()
```
changed to
```cs
.AddScoped<IBackendJwtHandler, ElsaBackendJwtHandler>()
```

When it was added as a singleton tokenAcquisition was always null. I tried using IServiceProvider to get the value but it was also always null.

Thank you for all the help. I appreciate it.

**Reply by zuntio** [reply]
Okay. That is strange. Work perfectly with singleton in my implementation. But in the end should not really matter since MemoryCache is Singleton underneath.

Thank you for taking the leap. Maybe I now have the courage to propose this generally as a solution

**Reply by laredoza** [reply]
@zuntio  Please propose this as a solution as it solves an important feature that is lacking. 

My POC is successful and we're going to go ahead with using Elsa. I'll provide feedback if I find anything.
